content/docs/command-reference/exp/apply.md

```diff
@@ -1,6 +1,6 @@
 # exp apply
 
-Apply the results from any [experiment](/doc/command-reference/exp) to the
+Apply the results from an [experiment](/doc/command-reference/exp) to the
 <abbr>workspace</abbr>.
```

Comment:
https://github.com/iterative/dvc.org/pull/2243#discussion_r586466572 makes me wonder if I got `exp apply` wrong. It does just basically checkout an experiment (which is already a "hidden commit") into the working tree right @pmrowla ?

And if so why do that and expect the user to (re)commit that when we could maybe just `exp commit` directly? Rebasing the refs/exps history onto `HEAD`

AND if so that would be very similar to `exp branch` so these could be merged probably (`exp commit --branch`)

Cc @dberenbaum 

Thanks

Reply:
For the record @jorgeorpinel, I had the exact same confusions as you, which is why I'm so concerned about how we document each of these.

`dvc exp apply` == `git merge --no-commit` (similar to `git stash apply` if we treat the experiment as a stash).

`dvc exp branch` == `git branch <branch> <exp_rev>` (similar to checking out the experiment, creating a new branch from it, then checking out the original branch again, except that it skips the checkouts).

These seem similar on the surface, but they differ in nonobvious ways. For one, `dvc exp apply` can only happen where the experiment is derived from the HEAD. Also, it requires stashing any uncommitted changes in the workspace and then trying to apply them on top of the applied experiment (and these may conflict, in which case `dvc exp apply` fails or ignores the stashed workspace changes depending on whether the `--no-force` flag is used). Finally, `dvc exp apply` ignores any previous checkpoints or other history associated with the experiment and only tries to apply the changes specified in the last experiment commit.

`dvc exp branch` seems more complex on the surface but is a simpler operation: it takes whatever history is tracked by that experiment and puts it into a new branch. It doesn't care what is in HEAD or the workspace since it's doing nothing more than adding <exp_rev> to `refs/heads/<branch>`.

Reply:
> And if so why do that and expect the user to (re)commit that when we could maybe just `exp commit` directly? Rebasing the refs/exps history onto `HEAD`

Users need a way to navigate to any experiment without adding it to their branch history, or else it basically defeats the purpose of experiments.

Reply:
> dvc exp apply == git merge --no-commit (similar to git stash apply if we treat the experiment as a stash).
>
> dvc exp branch == git branch <branch> <exp_rev> (similar to checking out the experiment, creating a new branch from it, then checking out the original branch again, except that it skips the checkouts).

This is correct with one exception:

>  Finally, dvc exp apply ignores any previous checkpoints or other history associated with the experiment and only tries to apply the changes specified in the last experiment commit.

This part is wrong, changes from the previous checkpoints are not ignored, they are also included in the merge. The exact equivalent for `dvc exp apply <rev>` would be `git merge --squash --no-commit <rev>`. All of the checkpoint commits are squashed into a single changeset, and then that is what gets merged into your head (as unstaged/uncommitted changes via `--no-commit`.

Reply:
Assuming that your current HEAD is the parent/baseline for a given experiment, the important difference between `exp branch ...` and `git checkout -b ...; exp apply ...; git commit` is that for checkpoints, using `exp apply` gives you a single `--squash`'ed commit in the end, whereas `exp branch` would include all of the individual checkpoint commits.

Reply:
Ah, sorry about that mistake. Should have left the explanation to Peter.

The point was to stress that there are a few subtle differences (how checkpoint commits are handled, how the current HEAD and workspace impacts the command) that came up as I was playing with experiments, and those differences seem really unintuitive without some understanding of the implementation.

I'll open a new issue to discuss how we should document experiment (and maybe I'll open similar issues for other new features if needed). We can gather feedback from users and decide whether we need to adjust anything there.

Reply:
Sorry to keep reviving this thread, but I only now realized that @pmrowla probably was intending `exp apply` and `exp branch` to mirror `git stash apply` and `git stash branch` (didn't realize before there was a `git stash branch`). So borrowing from Git docs is one idea. From https://git-scm.com/docs/git-stash, the justification for having `branch` in addition to `apply` is:

> This is useful if the branch on which you ran git stash push has changed enough that git stash apply fails due to conflicts. Since the stash entry is applied on top of the commit that was HEAD at the time git stash was run, it restores the originally stashed state with no conflicts.

Reply:
Hey, sorry for the delay here. OK sounds like I should def. review the current docs then... ⌛ 

But 💡 should we also consider letting users use Git commands instead of providing these complicated wrappers? And document recommended `git` operations instead (`merge/rebase --squash (--no-commit)`, `branch <exp-rev>` cherry-pick`, etc.)

> Users need a way to navigate to any experiment

Another example: `git checkout`

> @pmrowla probably was intending exp apply and exp branch to mirror git stash apply and git stash branch

Makes sense. Originally exps were going to piggy-back on the Git stash in fact.